backend/compact-connect/lambdas/staff-user-pre-token/user_scopes.py

```diff
@@ -0,0 +1,82 @@
+from config import config, logger
+
+
+class UserScopes(set):
+    """
+    Custom Set that will populate itself based on the user's database record contents
+    """
+    def __init__(self, sub: str):
+        super().__init__()
+        self._get_scopes_from_db(sub)
+
+    def _get_scopes_from_db(self, sub: str):
+        """
+        Parse the user's database record to calculate scopes.
+
+        Note: See the accompanying unit tests for expected db record shape.
+        :param sub: The `sub` field value from the Cognito Authorizer (which gets it from the JWT)
+        """
+        user_data = self._get_user_data(sub)
+        permissions = user_data.get('permissions', {})
+
+        # Ensure included compacts are limited to supported values
+        disallowed_compcats = permissions.keys() - config.compacts
+        if disallowed_compcats:
+            raise ValueError(f'User permissions include disallowed compacts: {disallowed_compcats}')
+
+        for compact_name, compact_permissions in permissions.items():
+            self._process_compact_permissions(compact_name, compact_permissions)
+
+    @staticmethod
+    def _get_user_data(sub: str):
+        try:
+            user_data = config.users_table.get_item(Key={'pk': sub})['Item']
+        except KeyError as e:
+            logger.error('Authenticated user not found!', exc_info=e, sub=sub)
+            raise RuntimeError('Authenticated user not found!') from e
+        return user_data
+
+    def _process_compact_permissions(self, compact_name, compact_permissions):
+        # Compact-level permissions
+        compact_actions = compact_permissions.get('actions', set())
```

Comment:
I don't see an `actions` key in the [diagram](https://github.com/csg-org/CompactConnect/blob/7f02302977dae3c7374e49ee0ea3cf07b1b047e9/backend/compact-connect/docs/design/users-arch-diagram.pdf)?

Reply:
This is true. The diagram is more of a conceptual model of how the permissions may be represented via API than an actual schema for the record in the database.